cairo/surface.c

```diff
@@ -126,12 +126,26 @@ static cairo_status_t
 _write_func (void *closure, const unsigned char *data, unsigned int length) {
   PyGILState_STATE gstate = PyGILState_Ensure();
   PyObject *res = PyObject_CallMethod ((PyObject *)closure, "write", "(" PYCAIRO_DATA_FORMAT "#)",
-				       data, (Py_ssize_t)length);
+                                       data, (Py_ssize_t)length);
   if (res == NULL) {
-    PyErr_Clear();
     /* an exception has occurred, it will be picked up later by
      * Pycairo_Check_Status()
+     *
+     * Do not clear the error here!
      */
+    PyObject *ptype, *pvalue, *ptraceback;
+    PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+
+    PyObject *error_string
+      = PyUnicode_FromFormat("Cairo: Exception occured while writing to output stream: %U",
+                             pvalue);
+    Py_XDECREF(ptype);
+    Py_XDECREF(pvalue);
+    Py_XDECREF(ptraceback);
+
+    PyErr_SetObject(PyExc_Exception, error_string);
```

Comment:
This will lead to the exception getting raised in unrelated code once things return to the interpreter.

Reply:
What do you mean by unrelated code? As soon as a Cairo function tries to write to the stream and fails, it returns because CAIRO_STATUS_WRITE_ERROR is returned. Then Pycairo_Check_Status runs and sees that an exception was raised. Hence, the function that caused the error to trigger will also raise the exception.

Reply:
After PyGILState_Release() any Python code can run (in another thread for example) and will fail because an exception is set.

Reply:
I see. I was not aware of this multi-threading issue. If I find some spare time, I will try to implement a way to properly pass the error message around, if you do not mind.